### PR TITLE
feat(diffusers/models): add models like autoencoders, transformers, unets and etc.

### PR DESCRIPTION
**Pull request by townwish4git**
# What does this PR do?

Adds # (feature)

implements of models in mindone.diffusers.models, include:

#### AutoEncoders

- [x] AutoencoderKL
- [x] AsymmetricAutoencoderKL
- [x] AutoencoderKLTemporalDecoder
- [x] ConsistencyDecoderVAE
- [x] AutoEncoderTiny
- [x] VQModel

#### UNets

- [ ] UNet1DModel
- [x] UNet2DModel
- [x] UNet2DConditionModel
- [x] UNet3DConditionModel
- [x] I2VGenXLUNet
- [x] Kandinsky3UNet
- [x] UNetSpatioTemporalConditionModel
- [x] UNetMotionModel
- [ ] StableCascadeUNet
- [x] UViT2DModel

#### Transformers

- [x] Transformer2DModel
- [x] TransformerTemporalModel
- [x] T5FilmDecoder
- [x] PriorTransformer
- [x] DualTransformer2DModel

supplement: https://gist.github.com/Cui-yshoho/7ff86a76323c37f1c197d7fef67702ec

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/mindspore-lab/mindone/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the documentation with your changes? E.g. record bug fixes or new features in `What's New`. Here are the
      [documentation guidelines](https://github.com/mindspore-lab/mindcv/wiki/%E6%96%87%E6%A1%A3%E7%BC%96%E5%86%99%E8%A7%84%E8%8C%83)
- [ ] Did you build and run the code without any errors?
- [ ] Did you report the running environment (NPU type/MS version) and performance in the doc? (better record it for data loading, model inference, or training tasks)
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@xxx

### CODE REVIEW

**Review by geniuspatrick — COMMENTED** [review]
yushi，huiyu合入后，rebase

**Comment by townwish4git** [comment]
> yushi，huiyu合入后，rebase， resolve conflict

done.

File: mindone/diffusers/README.md [hunk 1]
```diff
@@ -134,6 +150,39 @@ Most base, utility and mixin class are available.
 Unlike the output `posterior = DiagonalGaussianDistribution(latent)`, which can do sampling by `posterior.sample()`.
 We can only output the `latent` and then do sampling through `AutoencoderKL.diag_gauss_dist.sample(latent)`.
 
+### `nn.Conv3d`
```
[the remark this answers is not in the document]
Comment (geniuspatrick) [reply mindone/diffusers/README.md:153]: 之前说的不同精度和模式下的支持列表，先放在pr的说明中

[the remark this answers is not in the document]
Comment (geniuspatrick) [reply mindone/diffusers/README.md:153]: TODO: 后续考虑和yushi的列表和README中的limitation章节合并，提取出一个`Limitation.md`以作说明

File: mindone/diffusers/models/attention_processor.py [hunk 2]
```diff
@@ -430,7 +430,7 @@ def get_attention_scores(self, query: ms.Tensor, key: ms.Tensor, attention_mask:
             )
         else:
             attention_scores = ops.baddbmm(
-                attention_mask,
+                attention_mask.to(query.dtype),
```
Comment (geniuspatrick) [inline mindone/diffusers/models/attention_processor.py:433]: 这个为什么要加cast

File: mindone/diffusers/models/attention_processor.py [hunk 3]
```diff
@@ -475,7 +475,9 @@ def prepare_attention_mask(
             #       we want to instead pad by (0, remaining_length), where remaining_length is:
             #       remaining_length: int = target_length - current_length
             # TODO: re-enable tests/models/test_models_unet_2d_condition.py#test_model_xattn_padding
-            attention_mask = ops.pad(attention_mask, (0, target_length), value=0.0)
+            attention_mask = ops.Pad(paddings=((0, 0),) * (attention_mask.ndim - 1) + ((0, target_length),))(
```
Comment (geniuspatrick) [inline mindone/diffusers/models/attention_processor.py:478]: 需要手动`* (attention_mask.ndim - 1)`吗？

File: mindone/diffusers/models/autoencoders/vae.py [hunk 4]
```diff
@@ -306,6 +306,370 @@ def construct(
...
+            in_ch_ = out_ch_
+
+        # nn.SequentialCell does not support the len(self.layers) method nor does it support self.layers[l],
+        # therefore, we use nn.CellList instead.
+        self.layers = nn.CellList(layers)
+
+    def construct(self, x: ms.Tensor, mask=None) -> ms.Tensor:
+        r"""The forward method of the `MaskConditionEncoder` class."""
+        out = {}
+        for l in range(len(self.layers)):  # noqa: E741
```
Comment (geniuspatrick) [inline mindone/diffusers/models/autoencoders/vae.py:390]: why noqa?

File: mindone/diffusers/models/autoencoders/vae.py [hunk 5]
```diff
@@ -306,6 +306,370 @@ def construct(
...
+        legacy: bool = True,
+    ):
+        super().__init__()
+        self.n_e = n_e
+        self.vq_embed_dim = vq_embed_dim
+        self.beta = beta
+        self.legacy = legacy
+
+        self.embedding = nn.Embedding(self.n_e, self.vq_embed_dim)
+        self.embedding.embedding_table.set_data(
```
Comment (geniuspatrick) [inline mindone/diffusers/models/autoencoders/vae.py:573]: TODO: init method refactor

File: mindone/diffusers/models/transformers/t5_film_transformer.py [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+
+class NewGELUActivation(nn.Cell):
+    """
+    Implementation of the GELU activation function currently in Google BERT repo (identical to OpenAI GPT). Also see
+    the Gaussian Error Linear Units paper: https://arxiv.org/abs/1606.08415
+    """
+
+    def construct(self, input: ms.Tensor) -> ms.Tensor:
+        return (
+            0.5 * input * (1.0 + ops.tanh(float(math.sqrt(2.0 / math.pi)) * (input + 0.044715 * ops.pow(input, 3.0))))
```
Comment (geniuspatrick) [inline mindone/diffusers/models/transformers/t5_film_transformer.py:414]: `construct`里面使用`math`是否会有问题？考虑将常量手动放在外面

File: tests/diffusers/models/test_generic_modules.py
Comment (geniuspatrick) [inline tests/diffusers/models/test_generic_modules.py:1]: generic和transformers为什么分开？

File: mindone/diffusers/models/attention_processor.py (hunk 2 above)
Comment (townwish4git) [reply mindone/diffusers/models/attention_processor.py:433]: `ops.baddbmm`内部会检查`attention_mask`, `query`, `key`三者的dtype是否一致。在upcast_attention的情况下前者与后两者不一致，会报错

File: mindone/diffusers/models/attention_processor.py (hunk 3 above)
Comment (townwish4git) [reply mindone/diffusers/models/attention_processor.py:478]: `ops.Pad`需要`paddings`的shape与padded tensor的ndim对应（https://www.mindspore.cn/docs/zh-CN/r2.3.0rc1/api_python/ops/mindspore.ops.Pad.html?highlight=pad#mindspore.ops.Pad）

File: mindone/diffusers/models/autoencoders/vae.py (hunk 4 above)
Comment (townwish4git) [reply mindone/diffusers/models/autoencoders/vae.py:390]: 单独一个小字母的变量`l`不符合规范，为保持变量名和diffusers一致因此加了noqa

File: mindone/diffusers/models/transformers/t5_film_transformer.py (hunk 6 above)
Comment (townwish4git) [reply mindone/diffusers/models/transformers/t5_film_transformer.py:414]: Graph mode下对math.xxx()return结果的类型判断会存在问题，因此用`float()`包了一层让其能被判断为python内置类型。gitee mindspore某个issue提过这个问题和这个解法，具体链接找不到了

File: tests/diffusers/models/test_generic_modules.py
Comment (townwish4git) [reply tests/diffusers/models/test_generic_modules.py:1]:
```py
# In cases where models have unique initialization procedures or require testing with specialized output formats,
# it is necessary to develop distinct, dedicated test cases.
```

`t5_film_decoder`的输入输出格式不符合通用的测试用例结构，因此单独拉出来了。🤔可以考虑把t5的测试文件命名更清晰化，例如就叫`test_t5_film_decoder.py`？

File: mindone/diffusers/models/autoencoders/vae.py (hunk 5 above)
Comment (townwish4git) [reply mindone/diffusers/models/autoencoders/vae.py]: done, review plz

File: mindone/diffusers/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (townwish4git) [reply mindone/diffusers/README.md]: 已删除该部分。yushi的列表下方我也贴了`nn.Conv3d`[影响情况](https://gist.github.com/Cui-yshoho/7ff86a76323c37f1c197d7fef67702ec)的，我把它贴到这边

File: mindone/diffusers/models/autoencoders/vae.py [hunk 7]
```diff
@@ -306,6 +307,371 @@ def construct(
...
+        return sample
+
+
+class VectorQuantizer(nn.Cell):
+    """
+    Improved version over VectorQuantizer, can be used as a drop-in replacement. Mostly avoids costly matrix
+    multiplications and allows for post-hoc remapping of indices.
+    """
+
+    # NOTE: due to a bug the beta term was applied to the wrong term. for
```
Comment (CaitinZhao) [inline mindone/diffusers/models/autoencoders/vae.py:554]: what bug?

Comment (townwish4git) [reply mindone/diffusers/models/autoencoders/vae.py:554]: this comment comes from origin diffusers, not mindspore version 233

File: mindone/diffusers/models/transformers/t5_film_transformer.py (hunk 6 above)
Comment (CaitinZhao) [inline mindone/diffusers/models/transformers/t5_film_transformer.py:414]: 这块的type预期是什么？和input一致？动态图下这个的type推导可能有问题

Comment (townwish4git) [reply mindone/diffusers/models/transformers/t5_film_transformer.py:414]: > 这块的type预期是什么？和input一致？动态图下这个的type推导可能有问题

预期和input一致。在什么情况下会有问题呢？目前看起来这里的逻辑没啥问题？
```py
# python-built-in-float * ms.Tensor(dtype=...) => ms.Tensor(dtype=same...)
float(math.sqrt(2.0 / math.pi)) * (input + 0.044715 * ops.pow(input, 3.0))
```

Comment (townwish4git) [reply mindone/diffusers/models/transformers/t5_film_transformer.py:414]: 目前 {动态图， 静态图} x {fp16, fp32} 的单元测试都能跑通且和torch精度一致，或许bf16会有问题？之前确实遇到过 `ms.Tensor(bf16) * python-built-in-float => ms.Tensor(fp32)`的情况😂

**Comment by townwish4git** [comment]
use magic number `0.797885` to replace `math.sqrt(2.0 / math.pi)`, passed ut on cpu in {fp16, fp32} x {pynative, graph} with almost same results comparing to pytorch

**Review by CaitinZhao — APPROVED** [review]

**Pull request merged**